File: dart/algorithm/bin/custom_001.dart
import 'dart:io';
import 'dart:math';

void main() {
  int T = int.parse(stdin.readLineSync()!);

  for (int testNum = 0; testNum < T; testNum++) {
    var inputData = stdin.readLineSync()!.split(' ');
    var n = int.parse(inputData[0]), k = int.parse(inputData[1]);
    var interests =
        stdin.readLineSync()!.split(' ').map((s) => int.parse(s)).toList();

    stdout.writeln('#${testNum + 1}: ${solution(n, k, interests)}');
  }
}

int solution(int n, int k, List<int> interests) {
  var maxGroup = 0;
  var startIdx = 0;
  var endIdx = 0;

  interests.sort();

  while (endIdx < n) {
    while (endIdx < n && interests[endIdx] - interests[startIdx] <= k) {
      endIdx++;
    }

    maxGroup = max(maxGroup, endIdx - startIdx);

    while (endIdx < n && interests[endIdx] - interests[startIdx] > k) {
      startIdx++;
    }
  }

  return maxGroup;
}

File: python/Algorithm/CUSTOM001.py
def solution(N, K, interests):
    max_group = 0
    start_idx = 0
    end_idx = 0

    interests.sort()

    while end_idx < len(interests):
        while end_idx < len(interests) and interests[end_idx] - interests[start_idx] <= K:
            end_idx += 1

        max_group = max(max_group, interests[end_idx] - interests[start_idx])

        while end_idx < len(interests) and interests[end_idx] - interests[start_idx] > K:
            start_idx += 1

    return max_group


for _ in range(int(input())):
    N, K = list(map(int, input().split()))
    interests = list(map(int, input().split()))

    print('#%d: %d' % (_ + 1, solution(N, K, interests)))
